parsl/providers/ad_hoc/ad_hoc.py

```diff
@@ -0,0 +1,235 @@
+
+
+def _roundrobin(items):
+    """ Returns items one at a time in an infinite loop
+    """
+    while True:
+        for item in items:
+            yield item
+
+
+class AdHocProvider(ExecutionProvider, RepresentationMixin):
+    """ Ad-hoc execution provider
+
+    This provider is used to provision execution resources over one or more ad hoc nodes
+    that are each accessible over a Channel (say, ssh) but otherwise lack a cluster scheduler.
+
+    Each submit call invoked will go through the list of channels in a round-robin fashion
+
+    Parameters
+    ----------
+
+    channels : list of Channel ojects
+      Each channel represents a connection to a remote node
+
+    worker_init : str
+      Command to be run before starting a worker, such as 'module load Anaconda; source activate env'.
+      Since this provider calls the same worker_init across all nodes in the ad-hoc cluster, it is
+      recommended that a single script is made available across nodes such as ~/setup.sh that can
+      be invoked.
+
+    cmd_timeout : int
+      Duration for which the provider will wait for a command to be invoked on a remote system.
+      Defaults to 30s
+
+    parallelism : float
+      Determines the ratio of workers to tasks as managed by the strategy component
+
+    """
+
+    def __init__(self,
+                 channels=[],
+                 worker_init='',
+                 cmd_timeout=30,
+                 parallelism=1,
+                 move_files=None):
+
+        self.channels = channels
+        self._label = 'ad-hoc'
+        self.worker_init = worker_init
+        self.cmd_timeout = cmd_timeout
+        self.parallelism = 1
+        self.move_files = move_files
+        self.launcher = SimpleLauncher()
+        self.init_blocks = self.min_blocks = self.max_blocks = len(channels)
+
+        # This will be overridden by the DFK to the rundirs.
+        self.script_dir = "."
+
+        # In ad-hoc mode, nodes_per_block should be 1
+        self.nodes_per_block = 1
+
+        # Dictionary that keeps track of jobs, keyed on job_id
+        self.resources = {}
+
+        self.roundrobin = _roundrobin(self.channels)
+        logger.debug("AdHoc provider initialized")
+
+    def _write_submit_script(self, script_string, script_filename):
+        '''
+        Load the template string with config values and write the generated submit script to
+        a submit script file.
+
+        Args:
+              - template_string (string) : The template string to be used for the writing submit script
+              - script_filename (string) : Name of the submit script
+
+        Returns:
+              - None: on success
+
+        Raises:
+              SchedulerMissingArgs : If template is missing args
+              ScriptPathError : Unable to write submit script out
+        '''
+
+        try:
+            with open(script_filename, 'w') as f:
+                f.write(script_string)
+
+        except KeyError as e:
```

Comment:
when does this `KeyError` get raised?

Reply:
Looks like we don't do any command string formatting here anymore.